### PR TITLE
Add ability to specify GitHub events.

### PR DESCRIPTION
**Pull request by marcus-chan**
Adds functionality to support other GitHub event types other than push. Extension of https://github.com/jenkinsci/github-plugin/pull/53.

[Review on Reviewable](https://reviewable.io/reviews/jenkinsci/github-plugin/54)

### CODE REVIEW

**Comment by KostyaSha** [comment]
:-1: coding style changes, can't follow real changes. Plugin is not about any other event types rather then PUSH.

**Comment by jtillman** [comment]
:+1: PUSH event seems real limited. In my setup, we just build during Tagging/Branch Creation. This make sense for changes that go out in batch rather than individual items.

**Comment by KostyaSha** [comment]
This change has problems and useless complexity in every line of change, have no even enough time to comment them all. 

@jtillman please open and described your issue in jira.

**Comment by jtillman** [comment]
@KostyaSha I think my issues are already describe in an existing ticket: https://issues.jenkins-ci.org/browse/JENKINS-27842

**Comment by KostyaSha** [comment]
@jtillman about "webhooks" vs "services" i already answered in your PR. If push ignores tags, then it other issue. Don't mix many issues in one.

This PR wants add extra configurations and end user complexity without describing any real case. Allowing register many events in plugin that doesn't support this events is bad behaviour. In the same way gh plugin can do coffee.

**Comment by jtillman** [comment]
Sorry but I couldn't fully understand your last explanation. What I think that the ticket references is the ability to subscribe different jobs to different repositories and events? Briefly looking at the UI change in this PR, it seems it just adds an advanced configuration which doesn't effect the existing use case. I'd be willing to help clean it up if its just stylistic issues. I'd like to understand better your concern around having this more configurable.

**Comment by KostyaSha** [comment]
Events registered by this plugin incoming into this plugin, registering useless event that doesn't handled by this plugin is redundant configuration without any real case. Listener is the right direction but UI configuration in this plugin absolutely useless. 
Having variable for providing other variable optionally is also not good variant of UI configuration.
GitHubCause looks like a hack for his own single use case.  
I split current plugin issues in 2 issues https://issues.jenkins-ci.org/browse/JENKINS-28138 and https://issues.jenkins-ci.org/browse/JENKINS-28139 
And firstly i think should be resolved #39 , then migration to webhooks and finally extension.

File: pom.xml [hunk 1]
```diff
@@ -3,7 +3,7 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.532</version>
+    <version>1.610</version>
```
Comment (KostyaSha) [inline pom.xml]: Reason?

Comment (marcus-chan) [inline pom.xml]: Jenkins has a bug https://issues.jenkins-ci.org/browse/JENKINS-18537 that has problems with JDK8.

**Comment by adepue** [comment]
:+1:  Being able to wire to more than PUSH events would be nice for me as well.

File: pom.xml (hunk 1 above)
Comment (KostyaSha) [inline pom.xml]: Is it compile time issue of your java8 or runtime of all jenkins?

Comment (rogerhu) [inline pom.xml]: It's a runtime issue for Jenkins when using this plugin and running mvi hpi:run.  The moment you startup Jenkins with this plugin, you should be able to see the issue.

**Comment by rogerhu** [comment]
+1 -- allowing more granular GitHub event types makes sense and enabling this feature helps automate the process of installing webhooks across multiple repo's that only need to listen for certain events.

File: pom.xml (hunk 1 above)
Comment (oleg-nenashev) [inline pom.xml]: Makes sense to decrease the dependency to 1.609.1 when it gets released

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java [hunk 2]
```diff
@@ -53,8 +56,22 @@
  * @author Kohsuke Kawaguchi
  */
 public class GitHubPushTrigger extends Trigger<AbstractProject<?,?>> implements GitHubTrigger {
+
+    private boolean overrideEventTypes;
+    private EnumSet<GHEvent> eventTypes;
+
     @DataBoundConstructor
-    public GitHubPushTrigger() {
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: Binary compatibility loss. The original constructor should be retained somehow

File: src/main/java/com/cloudbees/jenkins/GitHubCause.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import java.io.IOException;
+
+/**
+ * UI object that says a build is started by GitHub post-commit hook.
+ *
+ * @author Marcus Chan
+ */
+public class GitHubCause extends SCMTriggerCause {
+
+    private String eventType;
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/GitHubCause.java]: I would explicitly mark fields and methods as <code>@CheckForNull</code>. There's inconsistency in the code: you default the value to empty string, but check against null in <code>getShortDescription</code>

File: src/main/java/com/cloudbees/jenkins/GitHubCause.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    private String name;
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/GitHubCause.java]: Makes sense to rename to "author"

File: src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java [hunk 5]
```diff
@@ -174,9 +200,37 @@ public void run() {
         });
     }
 
+    private void updateWebHook(GHRepository repo, URL url) {
+        try {
+            String urlExternalForm = url.toExternalForm();
+            GHHook hook = null;
+            for (GHHook h : repo.getHooks()) {
+                if (h.getName().equals("web") && h.getConfig().get("url").equals(urlExternalForm)) {
+                    hook = h;
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/GitHubPushTrigger.java]: and break?

File: src/main/java/com/cloudbees/jenkins/GitHubWebHook.java [hunk 6]
```diff
@@ -163,16 +164,23 @@ public void doIndex(StaplerRequest req, StaplerResponse rsp) {
...
+                    GHEvent.valueOf(eventType.toUpperCase());
+                    String payload = req.getParameter("payload");
+                    if (payload == null) {
+                        throw new IllegalArgumentException("Not intended to be browsed interactively (must specify payload parameter). " +
+                                "Make sure payload version is 'application/vnd.github+form'.");
+                    }
+
+                    processGitHubPayload(eventType, payload, GitHubPushTrigger.class);
+                } catch (IllegalArgumentException e) {
+                    throw new IllegalArgumentException("Github Webhook event of type " + eventType + " is not supported.");
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/GitHubWebHook.java]: Also catches the internal throw with a loss of data

File: src/main/java/com/cloudbees/jenkins/GitHubWebHook.java [hunk 7]
```diff
@@ -219,8 +232,13 @@ public void processGitHubPayload(String payload, Class<? extends Trigger<?>> tri
             } finally {
                 SecurityContextHolder.getContext().setAuthentication(old);
             }
-            for (Listener listener: Jenkins.getInstance().getExtensionList(Listener.class)) {
-                listener.onPushRepositoryChanged(pusherName, changedRepository);
+
+            for (Listener listener : Jenkins.getInstance().getExtensionList(Listener.class)) {
+                listener.onRepositoryChanged(eventType, name, changedRepository);
+                if ("push".equals(eventType)) {
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/GitHubWebHook.java]: should be checked before launching the for cycle

File: src/main/java/com/cloudbees/jenkins/GitHubWebHook.java [hunk 8]
```diff
@@ -254,7 +272,10 @@ public static GitHubWebHook get() {
          * @param changedRepository the changed repository.
          * @since 1.8
          */
+        @Deprecated
         public abstract void onPushRepositoryChanged(String pusherName, GitHubRepositoryName changedRepository);
+
+        public abstract void onRepositoryChanged(String eventType, String name, GitHubRepositoryName changedRepository);
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/jenkins/GitHubWebHook.java]: Javadoc is important for an extension point

**Comment by oleg-nenashev** [comment]
The PR cannot be merged in the current state. 
I suppose #57 and #60 supersede this PR

**Comment by rogerhu** [comment]
So are you OK with exposing event types in general that can be used for GitHub webhooks?

I've briefly looked at the new changes but wondering if you have any thoughts about the best way to integrate some of this stuff?

**Comment by KostyaSha** [comment]
@rogerhu you will need create separate plugin that will register extension points and consume anything you want. gh-plugin itself needs only pushes, pr plugin needs issue+pr, your plugin whatever event type you want.

**Comment by rogerhu** [comment]
Can the github-plugin be configured like the GitHub to be able to control what events to trigger?

The reason we like the GitHub plugin to do this it can manage all the webhooks related to Jenkins and GitHub integration.  We only need to trigger when there are builds/releases issuse and not for every push.

image

**Comment by KostyaSha** [comment]
Interesting idea, currently i'm working on new pullrequest plugin and it supports more triggering options. Probably we can try to do it more generic and call as github-trigger-plugin. 
@lanwen what do you think?

**Comment by rogerhu** [comment]
That would be great if you could add support... :)  That was the intent of this PR I believe... any branch that you currently working on?

**Comment by KostyaSha** [comment]
lanwen is working on credentials now, later we will think how better implement other triggers.

**Comment by lanwen** [comment]
yep, the next issue will be common gh trigger

**Comment by rogerhu** [comment]
Happy to help test...just point me to a branch when you're ready!

**Comment by KostyaSha** [comment]
@rogerhu we will notify you and request for testing. Atm we need prepare API for reusing them. For example credentials support.

File: src/main/java/com/cloudbees/jenkins/GitHubCause.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        super(pollingLog);
+        this.eventType = eventType;
+        this.name = name;
+    }
+
+    @Override
+    public String getShortDescription() {
+        String event = eventType != null ? eventType : "";
+        String author = name != null ? name : "";
+        return "Started by GitHub " + event + " by " + author;
```
Comment (shubhampa) [inline src/main/java/com/cloudbees/jenkins/GitHubCause.java]: Why constants, you can create constants at common constants file. use them here

**Pull request closed without merging**